### DISCUSSION TITLE
App pre close (cleanup) logic

### DISCUSSION THREAD

**Discussion by ixenion** [flet-dev/flet#3555] [Q&A]
Id like to do cleanup logic (call some functions) right before closing the flet app,
But I cant figure out how to do this.

There is "page.on_close" method, but seems it does something else
And not what I want. At least it does not fire when the app is closing.

**Comment by ndonkoHenri** [comment]
https://flet.dev/docs/reference/types/window#destroy

**Reply by ixenion** [reply]
Thanks, that would work (with some fixes) as desctop app.
But on android?.. Seems not.
Im sorry, thats my fault I wasnt clear enought, but is there solution for android?
I mean right before the app is closed (destroyed, removed from memory)

**Reply by ixenion** [reply]
I double checked. No, It definitely not working on android

**Reply by ndonkoHenri** [reply]
https://flet.dev/docs/controls/page/#on_app_lifecycle_state_change ?

**Reply by ixenion** [reply]
Seems it should work like so:

```python
import flet as ft


def main(page: ft.Page):
    page.title = "Flet counter example"
    page.vertical_alignment = ft.MainAxisAlignment.CENTER

    txt_number = ft.TextField(value="0", text_align=ft.TextAlign.RIGHT, width=100)
    number = page.client_storage.get("number")
    txt_number.value = number if number else "0"


    def minus_click(e):
        txt_number.value = str(int(txt_number.value) - 1)
        page.update()

    def plus_click(e):
        txt_number.value = str(int(txt_number.value) + 1)
        page.update()


    row = ft.Row(
            [
                ft.IconButton(ft.icons.REMOVE, on_click=minus_click),
                txt_number,
                ft.IconButton(ft.icons.ADD, on_click=plus_click),
            ],
            alignment=ft.MainAxisAlignment.CENTER,

        )

    page.add(
            row,
    )


    def handler(e:ft.AppLifecycleStateChangeEvent):
        if e.data == ft.AppLifecycleState.DETACH:
            page.client_storage.set("number", txt_number.value)
            page.window.destroy()

    page.on_app_lifecycle_state_change = handler
    page.window.prevent_close = True

ft.app(main)
```

It's standart counter template, but it should save count right before app destroyed.
And... It does not. It does not save count when I close and launch tha app again.
Maybe Im doing something wrong?

**Reply by Razielthc** [reply]
According to the documentation, the values ​​stored in the client will be deleted when you close the application, you should use a database or a file where you save the values ​​and recover them later https://flet.dev/docs/cookbook/client-storage

**Reply by ixenion** [reply]
> According to the documentation, the values ​​stored in the client will be deleted when you close the application, you should use a database or a file where you save the values ​​and recover them later https://flet.dev/docs/cookbook/client-storage

Could you show exact sentence where it says "the values will be deleted on close" ?
Because I only see in your link "allows storing key-value data on a client side in a persistent storage." which is opposite to your state.
Also when Im using windows app written in flet there is no problem to store in client storage. The values persist between app restart.

**Reply by ixenion** [reply]
> According to the documentation, the values ​​stored in the client will be deleted when you close the application, you should use a database or a file where you save the values ​​and recover them later https://flet.dev/docs/cookbook/client-storage

What are you talking about!?
I can successfully store value if I make special button to store it. And it persists even after close/open cycle on android.
